In [10]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

In [11]:
# Train Phase transformations
train_transforms = transforms.Compose([
                                      #  transforms.Resize((28, 28)),
                                      #  transforms.ColorJitter(brightness=0.10, contrast=0.1, saturation=0.10, hue=0.1),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,)) # The mean and std have to be sequences (e.g., tuples), therefore you should add a comma after the values. 
                                       # Note the difference between (0.1307) and (0.1307,)
                                       ])

# Test Phase transformations
test_transforms = transforms.Compose([
                                      #  transforms.Resize((28, 28)),
                                      #  transforms.ColorJitter(brightness=0.10, contrast=0.1, saturation=0.10, hue=0.1),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))
                                       ])


In [12]:
train = datasets.MNIST('./data', train=True, download=True, transform=train_transforms)
test = datasets.MNIST('./data', train=False, download=True, transform=test_transforms)

In [13]:
SEED = 1

# CUDA?
cuda = torch.cuda.is_available()
print("CUDA Available?", cuda)

# For reproducibility
torch.manual_seed(SEED)

if cuda:
    torch.cuda.manual_seed(SEED)

# dataloader arguments - something you'll fetch these from cmdprmt
dataloader_args = dict(shuffle=True, batch_size=128, num_workers=4, pin_memory=True) if cuda else dict(shuffle=True, batch_size=64)

# train dataloader
train_loader = torch.utils.data.DataLoader(train, **dataloader_args)

# test dataloader
test_loader = torch.utils.data.DataLoader(test, **dataloader_args)

CUDA Available? True


/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


In [14]:

class Net(nn.Module):

  def __init__(self):
    super(Net, self).__init__()
    
    #convolutuonla block 1
    self.convblock1 = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=8, kernel_size=(3, 3), padding=0, bias=False),
            nn.ReLU()
    
    ) #input_size = 28, #output_size = 26, RF = 3
    self.convblock2 = nn.Sequential(
            nn.Conv2d(in_channels=8, out_channels=16, kernel_size=(3, 3), padding=0, bias=False),
            nn.ReLU()

    ) #input_size = 26, #output_size = 24, RF = 5
    """
    self.convblock3 = nn.Sequential(
            nn.Conv2d(in_channels=20, out_channels=20, kernel_size=(3, 3), padding=0, bias=False),
            nn.BatchNorm2d(20),
            nn.ReLU(),
            nn.Dropout(drop_out)
    ) #output_size = 22
   """
    # transition block
    self.maxpool1 = nn.MaxPool2d(2,2) #input_size = 24, #output_size = 12, RF = 6
    self.convblock4 = nn.Sequential(
            nn.Conv2d(in_channels=16, out_channels=8, kernel_size=(1, 1), padding=0, bias=False),
            nn.ReLU()
           
    )#input_size = 12, #output_size = 10, RF = 10

    #convolutuonla block 2
    self.convblock5 = nn.Sequential(
            nn.Conv2d(in_channels=8, out_channels=8, kernel_size=(3, 3), padding=0, bias=False),
            nn.ReLU()
        
    )#input_size = 10, #output_size = 8, RF = 14
    self.convblock6 = nn.Sequential(
            nn.Conv2d(in_channels=8, out_channels=16, kernel_size=(3, 3), padding=0, bias=False),
            nn.ReLU()
    )#input_size = 8, #output_size = 6, RF = 18
    self.convblock7 = nn.Sequential(
            nn.Conv2d(in_channels=16, out_channels=16, kernel_size=(3, 3), padding=0, bias=False),
            nn.ReLU()
    )#input_size = 6, #output_size = 4, RF = 22
    self.convblock8 = nn.Sequential(
            nn.Conv2d(in_channels=16, out_channels=16, kernel_size=(3, 3), padding=0, bias=False),
            nn.ReLU()
        
    )#input_size = 4, #output_size = 2, RF = 26

    #output block
    self.gap = nn.AdaptiveAvgPool2d((1,1)) #input_size = 2, #output_size = 1, RF = 28
    self.convblock9 = nn.Sequential(
            nn.Conv2d(in_channels=16, out_channels=10, kernel_size=(1,1), padding=0, bias=False),
            
    )

  def forward(self,input):

    #x = self.convblock3(self.convblock2(self.convblock1(input)))
    x = self.convblock2(self.convblock1(input))

    x = self.convblock4(self.maxpool1(x))

    x = self.convblock8(self.convblock7(self.convblock6(self.convblock5(x))))

    x = self.convblock9(self.gap(x))

    return F.log_softmax(x)



In [15]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
print(device)
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

cuda
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
            Conv2d-3           [-1, 16, 24, 24]           1,152
              ReLU-4           [-1, 16, 24, 24]               0
         MaxPool2d-5           [-1, 16, 12, 12]               0
            Conv2d-6            [-1, 8, 12, 12]             128
              ReLU-7            [-1, 8, 12, 12]               0
            Conv2d-8            [-1, 8, 10, 10]             576
              ReLU-9            [-1, 8, 10, 10]               0
           Conv2d-10             [-1, 16, 8, 8]           1,152
             ReLU-11             [-1, 16, 8, 8]               0
           Conv2d-12             [-1, 16, 6, 6]           2,304
             ReLU-13             [-1, 16, 6, 6]               0
           Conv2d-14             [

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:72: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


In [16]:
from tqdm import tqdm

train_losses = []
test_losses = []
train_acc = []
test_acc = []

def train(model, device, train_loader, optimizer, epoch):
  model.train()
  #pbar = tqdm(train_loader)
  correct = 0
  processed = 0
  for batch_idx, (data, target) in enumerate(train_loader):
    # get samples
    data, target = data.to(device), target.to(device)
    #print("target size:",target.size())
    #target = target.squeeze(1)
    #print("target size2:",target.size())
    # Init
    optimizer.zero_grad()
    # In PyTorch, we need to set the gradients to zero before starting to do backpropragation because PyTorch accumulates the gradients on subsequent backward passes. 
    # Because of this, when you start your training loop, ideally you should zero out the gradients so that you do the parameter update correctly.

    # Predict
    y_pred = model(data)
    #print("y_pred size:",y_pred.size())
    y_pred = y_pred.squeeze(-1)
    y_pred = y_pred.squeeze(-1)
    #print("y_pred size2:",y_pred.size())
    # Calculate loss
    loss = F.nll_loss(y_pred, target)
    train_losses.append(loss)

    # Backpropagation
    loss.backward()
    optimizer.step()

    # Update pbar-tqdm
    
    pred = y_pred.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
    correct += pred.eq(target.view_as(pred)).sum().item()
    processed += len(data)

    #pbar.set_description(desc= f'Loss={loss.item()} Batch_id={batch_idx} Accuracy={100*correct/processed:0.2f}')
    train_acc.append(100*correct/processed)
  return (100*correct/processed)

def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            output = output.squeeze(-1)
            output = output.squeeze(-1)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    test_losses.append(test_loss)
    
    """
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))
    """
    test_acc.append(100. * correct / len(test_loader.dataset))
    return (100. * correct / len(test_loader.dataset))

In [18]:
model =  Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
EPOCHS = 14
for epoch in range(EPOCHS):
    #print("EPOCH:", epoch)
    train_accuracy = train(model, device, train_loader, optimizer, epoch)
    test_accuracy = test(model, device, test_loader)
    print("epoch no: ",epoch," train_accuracy: ",train_accuracy," test_accuracy: ",test_accuracy)

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:72: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


epoch no:  0  train_accuracy:  9.301666666666666  test_accuracy:  10.06
epoch no:  1  train_accuracy:  10.531666666666666  test_accuracy:  12.97
epoch no:  2  train_accuracy:  28.546666666666667  test_accuracy:  86.03
epoch no:  3  train_accuracy:  91.89166666666667  test_accuracy:  95.59
epoch no:  4  train_accuracy:  95.8  test_accuracy:  96.06
epoch no:  5  train_accuracy:  96.91333333333333  test_accuracy:  97.47
epoch no:  6  train_accuracy:  97.38833333333334  test_accuracy:  98.02
epoch no:  7  train_accuracy:  97.73166666666667  test_accuracy:  98.24
epoch no:  8  train_accuracy:  97.93166666666667  test_accuracy:  98.34
epoch no:  9  train_accuracy:  98.14833333333333  test_accuracy:  98.32
epoch no:  10  train_accuracy:  98.29833333333333  test_accuracy:  97.92
epoch no:  11  train_accuracy:  98.46833333333333  test_accuracy:  98.31
epoch no:  12  train_accuracy:  98.52666666666667  test_accuracy:  98.44
epoch no:  13  train_accuracy:  98.59  test_accuracy:  98.26


Target :

    Bringing the no.of parameters around 8k (required for late students)
    Introduction of GAP before output block to reduce parameters
    

Results :

    Parameters - 8,024
    Best Train accuracy - 98.59 (within 14 epochs)
    Best Test accuracy - 98.44

Analysis :

    Less overfitting in model
    Accuracy is less as the no.of parameters reduced a lot
    Model performance can be improved by introducing regularisation techniques